### DISCUSSION TITLE
xrepo: Handling upstream binaries

### DISCUSSION THREAD

**Discussion by barcharcraz** [xmake-io/xmake#3045] [Ideas]
Some of the larger xrepo package scripts, such as llvm and Qt, download and install binaries from "upstream" instead of building from source. This is usually not a bad idea, as these projects take a long time to build, and may need very specific compiler flags, or involve very annoying to build dependencies (Qt webengine comes to mind).

Still, sometimes it would be nice to build from source instead.

I'm not sure how this should be controlled exactly, there's already the package.precompiled policy but that's for getting precompiled builds that correspond to xrepo build scripts. Additionally it's not completely unreasonable to make a "precompiled" package derived from a script that extracts upstream binaries.

**Comment by xq114** [comment]
You can add an option to those packages:
```
package("xxx")
	...
	add_configs("source", {description = "Build from source.", default = false, type = "boolean"})
	on_load(... function (package)
		if package:config("source") then
			package:set("urls", {"url/to/source"})
		end
	end)

	on_install(... function (package)
		if package:config("source") then
			...
		else
			...
		end
	end)
```

**Reply by barcharcraz** [reply]
It's a little annoying all the stuff to add URLs completely changes.

Is there a shortcut like `on_install("@linux", ...)` for this?

**Reply by waruqi** [reply]
you can also define your llvm package to build source code in your project xmake.lua.
